Add upstream strategy `strict`

DerRockWolf opened the pull request:
This PR contains a lot of changes (sorry for that :sweat_smile:)
and closes:
- #355
- #1001 

#### Summary of changes:
- Add `upstreams.strategy` field
- Add `strict` upstream resolver (with tests & docs)
- Refactor upstream resolver creation to be "tree-like" -> `upstreamTreeResolver` (as suggested by @ThinkChaos in #1001)
- Refactor: always return `Chained(Resolver)` and use type assertion in tests if needed

#### Details regarding the `upstremTreeResolver` and the "select group by client"

The `upstreamTreeResolver` contains branches (`map[string]Resolver`), each branch represents a groups upstream resolver. The `upstreamTreeResolver` just calls `Resolve` of the respective branch/group upstream resolver determined by the `upstreamGroupByClient` func.

Currently a client can "belong" to multiple groups and therefore would use all configured upstream servers (including possible duplicates) where the group name is matching clientName, IP or CIDR.
This behavior is IMO is really confusing and also not really documented or ensured via tests.

I've therefore replaced the "select upstream servers by client" to a "select group by client". It follows a strict order (currently: 1. exact IP, 2. CIDR, 3. clientNames). This would maybe be a breaking change for setups where clients are currently matching multiple groups and would use servers from the matching groups.

What do you think about my proposed change?

Additionally if someone would use e.g. two CIDR groups where an IP matches both, we can't ensure that really the first (as specified in the config file) is used as the `config.upstreams.groups` field is a map. Same applies for `clientNames`.

---

Do you want me to also extend the `server.go` and/or e2e tests? If so please tell me where and to what extend.

TODOs:
- [x] Clarify "select group by client" topic

0xERR0R commented on resolver/upstream_tree_resolver_test.go:
```diff
@@ -0,0 +1,281 @@
+					return res
+				}
+
+				clientSpecificExactMock := &mockResolver{}
+				clientSpecificExactMock.On("Resolve", mock.Anything)
+
+				clientSpecificWildcardMock := &mockResolver{}
+				clientSpecificWildcardMock.On("Resolve", mock.Anything)
+
+				clientSpecificIPMock := &mockResolver{}
+				clientSpecificIPMock.On("Resolve", mock.Anything)
+
+				clientSpecificCIRDMock := &mockResolver{}
+				clientSpecificCIRDMock.On("Resolve", mock.Anything)
```
Tried to understand this test... It looks like these mock resolvers are not used, right?

DerRockWolf replied:
You're correct. I forgot to remove them after adding the `createMockResolver` func :sweat_smile:

0xERR0R reviewed (changes requested):
I skimmed the code changes and it looks pretty good so far. :+1: 
I didn't test it yet, but I'll do it next time. Please check my comments :handshake:

0xERR0R commented:
> I've therefore replaced the "select upstream servers by client" to a "select group by client". It follows a strict order (currently: 1. exact IP, 2. CIDR, 3. clientNames). This would maybe be a breaking change for setups where clients are currently matching multiple groups and would use servers from the matching groups.

This sounds reasonable to me. We should document the order in the documentation to avoid confusions.

DerRockWolf commented on resolver/blocking_resolver.go:
```diff
@@ -95,7 +95,7 @@ type BlockingResolver struct {
 func NewBlockingResolver(
 	cfg config.BlockingConfig, redis *redis.Client, bootstrap *Bootstrap,
-) (r *BlockingResolver, err error) {
+) (r ChainedResolver, err error) {
```
I'm fine with always returning the struct pointer.

I've changed all resolver to return the interface type, because I was confused what my new resolver should return, as some of the existing resolvers were returning a struct pointer and others the interface.
I didn't know about the "Accept interfaces, return structs" pattern until now, thanks :slightly_smiling_face: 

If you want I can refactor all resolvers to always return its struct pointer.

0xERR0R replied:
Yes, please change the return type to struct and remove casting in tests :+1:

DerRockWolf commented:
> > I've therefore replaced the "select upstream servers by client" to a "select group by client". It follows a strict order (currently: 1. exact IP, 2. CIDR, 3. clientNames). This would maybe be a breaking change for setups where clients are currently matching multiple groups and would use servers from the matching groups.
> 
> This sounds reasonable to me. We should document the order in the documentation to avoid confusions.

Great. I'll add it to the docs :+1: 

How should we handle my last concern regarding this:

> Additionally if someone would use e.g. two CIDR groups where an IP matches both, we can't ensure that really the first (as specified in the config file) is used as the `config.upstreams.groups` field is a map. Same applies for `clientNames` groups.

For now, I think it would be sufficient to add this to the docs. In the future we could validate the config and check for colliding CIDRs & clientNames at startup. WDYT?

kwitsch commented:
I think it would also be reasonable to also log it as a warning.

It would speed up support cases where users report unexpected behavior(in my opinion).
If we know that it could happen, let them know. 😊

DerRockWolf commented:
I've added the order and further explanation to the configuration docs @0xERR0R
and added a log to warn users if a client matches multiple groups @kwitsch

0xERR0R commented:
I tested different configs and it worked fine :+1: 

Please resolve open points from discussions above.

@kwitsch Do you have any comments regarding this implementation?

kwitsch commented:
> I tested different configs and it worked fine :+1: 
> 
> Please resolve open points from discussions above.
> 
> @kwitsch Do you have any comments regarding this implementation?

I'm currently unable to test it but I've read it and it looks ok so far.

My only cretic would be the change of return types from struct pointers to interfaces as you also mentioned.
It's a bit confusing to read but that could only be me...😅

ThinkChaos commented:
Haven't been able to look at this since I don't have any computer time ATM and for a couple more weeks.

> It follows a strict order (currently: 1. exact IP, 2. CIDR, 3. clientNames)

Instinctively I'd expect the client names to override CIDR since it seems more specific, so could be used to override a CIDR.  
I don't use theses features though, so you probably know best, just thought I'd give a small bit of input.

kwitsch commented:
> Haven't been able to look at this since I don't have any computer time ATM and for a couple more weeks.

Same as me. Without the GitHub App on my phone I couldn't participate at all right now... 

> > It follows a strict order (currently: 1. exact IP, 2. CIDR, 3. clientNames)
> 
> Instinctively I'd expect the client names to override CIDR since it seems more specific, so could be used to override a CIDR.  
> I don't use theses features though, so you probably know best, just thought I'd give a small bit of input.

Would have said the same but I also don't use it. 🤷‍♂️

0xERR0R commented:
Makes sense. Client names should have higher priority. 👍

kwitsch reviewed (changes requested):
Open changes as discussed:
- [x ] Change return types to struct pointers 
- [ ] Hostnames should have priority over CIDR

DerRockWolf commented:
I've refactored every resolver (except upstream_tree & rewriter as they can also return a different type) to return a struct pointer
and changed the order of the "select group by client" logic (also updated the docs). :slightly_smiling_face:

kwitsch reviewed (approved):
I'm still unable to test it but I've read the code and cross checked against the previous behavior.

The changes are reasonable, documented and covered with unit tests.

👍

0xERR0R commented:
@DerRockWolf: Great job, thank you!
